File: loginClient.js
// CWE-319: Cleartext Transmission of Sensitive Information
//
// Credentials and a session token are sent over plain HTTP, exposing them to
// anyone able to observe network traffic. The fix is to use https.

const http = require("http");

function authenticate(username, password, apiToken) {
  const payload = JSON.stringify({ username, password });

  const options = {
    hostname: "api.example.com",
    port: 80,
    path: "/v1/login",
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      // CWE-319: API token sent in cleartext over HTTP
      Authorization: `Bearer ${apiToken}`,
      "Content-Length": Buffer.byteLength(payload),
    },
  };

  // CWE-319: using the unencrypted http module instead of https
  const req = http.request(options, (res) => {
    console.log(`HTTP ${res.statusCode}`);
  });

  req.write(payload);
  req.end();
}

authenticate("alice", "S3cr3tP@ss", "tok_live_abc123");

module.exports = { authenticate };

File: login_client.py
"""CWE-319: Cleartext Transmission of Sensitive Information.

Credentials and an API token are sent over plain HTTP, so they can be read by
anyone on the network path. The fix is to use an https:// URL (TLS).
"""

import urllib.request
import json


def authenticate(username: str, password: str, api_token: str) -> int:
    # CWE-319: sensitive data sent over an unencrypted HTTP connection
    url = "http://api.example.com/v1/login"
    payload = json.dumps({"username": username, "password": password}).encode("utf-8")

    request = urllib.request.Request(url, data=payload, method="POST")
    request.add_header("Content-Type", "application/json")
    # API token transmitted in cleartext header
    request.add_header("Authorization", f"Bearer {api_token}")

    with urllib.request.urlopen(request) as response:
        return response.status


if __name__ == "__main__":
    authenticate("alice", "S3cr3tP@ss", "tok_live_abc123")
